### PR TITLE
rustc_codegen_ssa: Better code generation for niche discriminants.

### PR DESCRIPTION
**Pull request by mikebenfield**
In some cases we can avoid arithmetic before checking whether a niche is a tag.

Also rename some identifiers around niches.

This is relevant to #101872

### CODE REVIEW

**Comment by mikebenfield** [comment]
In my performance testing this improves runtime of every benchmark I tried from `rustc-perf`. Usually around 1% or 2%, sometimes as much as 5%. However I often see discrepancies between the `perf` runs here and my local runs, so who knows. 

I felt some of the names around niches were confusing so I renamed some things. If it's too obnoxious of me to include the renames here I can change them back.

File: compiler/rustc_const_eval/src/interpret/operand.rs [hunk 1]
```diff
@@ -650,7 +650,7 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         // declared list of variants -- they can differ with explicitly assigned discriminants.
         // We use "tag" to refer to how the discriminant is encoded in memory, which can be either
         // straight-forward (`TagEncoding::Direct`) or with a niche (`TagEncoding::Niche`).
-        let (tag_scalar_layout, tag_encoding, tag_field) = match op.layout.variants {
+        let (tag_scalar_layout, encoding, field) = match op.layout.variants {
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/operand.rs:653]: Why did you rename these? The names were carefully chosen. Is 'tag' no longer accurate?

File: compiler/rustc_const_eval/src/interpret/operand.rs [hunk 2]
```diff
@@ -666,8 +666,8 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                 };
                 return Ok((discr, index));
             }
-            Variants::Multiple { tag, ref tag_encoding, tag_field, .. } => {
-                (tag, tag_encoding, tag_field)
+            Variants::Multiple { scalar, ref encoding, field, .. } => {
+                (scalar, encoding, field)
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/operand.rs:670]: 'scalar', in the interpreter, usually refers to an `interpret::Scalar`. Please avoid using that name for things that have another type.

File: compiler/rustc_const_eval/src/interpret/validity.rs [hunk 3]
```diff
@@ -220,8 +220,8 @@ impl<'rt, 'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> ValidityVisitor<'rt, 'mir, '
     fn aggregate_field_path_elem(&mut self, layout: TyAndLayout<'tcx>, field: usize) -> PathElem {
         // First, check if we are projecting to a variant.
         match layout.variants {
-            Variants::Multiple { tag_field, .. } => {
-                if tag_field == field {
+            Variants::Multiple { field, .. } => {
+                if field == field {
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/validity.rs:224]: Something went very wrong here.

File: compiler/rustc_target/src/abi/mod.rs [hunk 4]
```diff
@@ -1121,38 +1121,44 @@ pub enum Variants<'a> {
...
+    /// assign explicit discriminant values). That discriminant is encoded
+    /// in a field of the object. The layout of each variant is
+    /// a struct. For enums, the tag is the sole field of the layout.
     Multiple {
-        tag: Scalar,
-        tag_encoding: TagEncoding,
-        tag_field: usize,
+        encoding: TagEncoding,
+        scalar: Scalar,
+        field: usize,
```
Comment (RalfJung) [inline compiler/rustc_target/src/abi/mod.rs:1130]: I could live with removing the `tag_` prefix (but it seems you did a global search-and-replace here and that changed the meaning of some code, see my previous comment), but `scalar: Scalar` is worse than before IMO. This is the scalar ABI of the tag.

You also removed the term 'tag' from the comment, which is bad IMO. We used to have a lot of confusion between discriminant and tag, until I started systematically putting these terms everywhere. Please don't revert us back to the confusing state.

File: compiler/rustc_codegen_ssa/src/mir/place.rs [hunk 5]
```diff
@@ -222,88 +224,121 @@ impl<'a, 'tcx, V: CodegenObject> PlaceRef<'tcx, V> {
...
+
+                let lo = cmp::min(untagged_variant.as_u32(), tagged_variants.start().as_u32());
+                let hi = cmp::max(untagged_variant.as_u32(), tagged_variants.end().as_u32());
+                let lo = bx.cx().const_uint(cast_to, lo as u64);
+                let hi = bx.cx().const_uint(cast_to, hi as u64);
+                let lo_cmp = bx.icmp(IntPredicate::IntULE, lo, value);
+                let hi_cmp = bx.icmp(IntPredicate::IntULE, value, hi);
+                bx.assume(lo_cmp);
+                bx.assume(hi_cmp);
+                value
```
Comment (bjorn3) [inline compiler/rustc_codegen_ssa/src/mir/place.rs:341]: Did you forget to apply most of this extra logic to cg_clif?

**Comment by RalfJung** [comment]
I can't comment on the codegen change, but the rename overall is a net negative IMO. I am open to suggestions for how to improve our current names, but we need a term for 'the discriminant in its encoded form in memory'. That thing used to be called 'tag', and this PR leaves that term in some places (`TagEncoding`) but removes it in many others (in particular in many of the comments) without even introducing an alternative term. If you want to replace 'tag' by something else, fine, but we need *some* way to refer to this, and this PR seems to have the goal of just eliminating the term without a replacement and without eliminating the underlying concept.

File: compiler/rustc_const_eval/src/interpret/operand.rs (hunk 2 above)
Comment (mikebenfield) [reply compiler/rustc_const_eval/src/interpret/operand.rs:670]: Alright. But given that objects of this type shouldn't be referred to as `scalar`s... maybe `Scalar` is not a good name for the type?

Comment (RalfJung) [reply compiler/rustc_const_eval/src/interpret/operand.rs:670]: I consider its full name to be `abi::Scalar`.

My understanding is that the name makes more sense outside of the interpreter, but given that the interpreter has `interpret::Scalar` we have a somewhat unfortunate conflict.

**Comment by RalfJung** [comment]
As I said I am open to renaming 'tag' if you have some good ideas. But I don't like just removing the name without replacement -- concepts like that need names. I'm not saying our current terms are great, they are not, but even bad terms are better than no terms.

The way I think about it: "niche" values of a type are bit patterns that are not valid representations for that type, and can be used to store the tag which can then be mapped to the discriminant. (In that sense they are not even values...)

**Comment by krdln** [comment]
@mikebenfield Looks like we've both noticed the "needless sub before cmp" and implemented basically the same idea, but you pushed your PR first :laughing:. I actually abandoned my branch, as perf wins I was seeing were very minuscule (despite asm looking quite nice in some artificial tests (see [gist])). I've ran your branch through this gist (see `-mike` file) and interestingly, it regresses in some cases (the `bar` / `baz` etc. functions, which are basically `matches!` on a single variant). The "full-match" (or calling `std::mem::get_discriminant`) is a very nice win though (I think I have a slight regression there)! Anyway, decided to undust and [push my code as a draft](https://github.com/rust-lang/rust/pull/102901).

So, on my branch there are some nice assembly improvements, but almost negligible perf wins, while you have some assembly regressions (and couple improvements) but significant perf wins. This is interesting! Perhaps the case in which I have slight regression, and you have an improvement – a "full-match on u8-based enum", is dominating? I hope we can analyze it further and come up with "best-of-both-worlds" approach.

[gist]: https://gist.github.com/krdln/cefbb4a14d7ddfb5014099d1a0060916

**Comment by mikebenfield** [comment]
Oh that's funny we both had the same idea. I didn't check on cases like your `bar` and `baz`; I'm surprised my code regresses in that case. Are you sure your `assume` is always true? `niche_variants` doesn't always have a hole.

**Comment by krdln** [comment]
@mikebenfield 
> Are you sure your assume is always true? niche_variants doesn't always have a hole.

You're right. I did a bit of mental shortcut – what I've meant is that `niche_variants` won't ever contain `tagged_variant` (if `tagged_variant` is outside `niche_variants`, this assumption while less useful, will still be valid. I even tried to gate the assume to "true holes only", but it's not worth it).

**Comment by mikebenfield** [comment]
OK, here appears to be the reason my codegen regresses in that case. 

Given this LLVM IR:
```LLVM
define i1 @myfunction(i8 %0) {
start:
  %1 = icmp ule i8 %0, 49
  %2 = zext i8 %0 to i64
  %3 = add i64 %2, -46
  %_2 = select i1 %1, i64 %3, i64 0
  %4 = icmp eq i64 %_2, 1
  ret i1 %4
}

```
(notice that `%z` basically is `%0`)
I run an `InstCombinePass` on it: (`opt --instcombine -S nope.ll`)
```LLVM
define i1 @myfunction(i8 %0) {
start:
  %1 = icmp eq i8 %0, 47
  ret i1 %1
}
```

Great! That works as expected. But instead change the function to take an `i16`, and truncate it:
```LLVM
define i1 @myfunction(i16 %0) {
start:
  %z = trunc i16 %0 to i8
  %1 = icmp ule i8 %z, 49
  %2 = zext i8 %z to i64
  %3 = add i64 %2, -46
  %_2 = select i1 %1, i64 %3, i64 0
  %4 = icmp eq i64 %_2, 1
  ret i1 %4
}
```

and the `InstCombinePass` produces this:

```LLVM
define i1 @myfunction(i16 %0) {
start:
  %z = trunc i16 %0 to i8
  %1 = icmp ult i8 %z, 50
  %z.mask = and i16 %0, 255
  %2 = icmp eq i16 %z.mask, 47
  %3 = and i1 %1, %2
  ret i1 %3
}
```

Which is... not so great.

**Comment by mikebenfield** [comment]
I filed [this issue](https://github.com/llvm/llvm-project/issues/58275).

**Comment by mikebenfield** [comment]
> a "full-match on u8-based enum" – is dominating?

Yeah most of my testing was on matches that are full or nearly full, generally with enough cases so that a jump table is generated.

File: compiler/rustc_codegen_ssa/src/mir/place.rs (hunk 5 above)
Comment (mikebenfield) [reply compiler/rustc_codegen_ssa/src/mir/place.rs]: I guess cf_clif refers to cranelift. Yeah I made no changes to code generation for cranelift.

**Comment by mikebenfield** [comment]
Cleaned up `codegen_get_discr` a little, wrote better comments, and made it more careful about when it can move all the arithmetic to after the cast.

Add test `src/test/ui/enum-discriminant/get_discr.rs` to make sure some of those cases work correctly.

**Comment by nagisa** [comment]
Before I take a look at this, quantifying the wins is probably a good way to get an objective answer to performance discussions above.

@bors try @rust-timer queue

**Comment by mikebenfield** [comment]
Cool, although note that currently [the LLVM issue I filed](https://github.com/llvm/llvm-project/issues/58275) is probably having a substantial negative impact.

**Comment by krdln** [comment]
@mikebenfield Looks like you still do have regressions on the `is_x()`-style functions. Perhaps this is the contributor to the regressions? Check out this gist https://gist.github.com/krdln/262f932d8ae75f19673bae3b2165f1c1.

Another hypothesis – perhaps too much of assumes do hurt? https://llvm.org/docs/Frontend/PerformanceTips.html

> Avoid using the [assume intrinsic](https://llvm.org/docs/LangRef.html#int-assume) until you’ve established that a) there’s no other way to express the given fact and b) that fact is critical for optimization purposes. Assumes are a great prototyping mechanism, but they can have negative effects on both compile time and optimization effectiveness. The former is fixable with enough effort, but the later is fairly fundamental to their designed purpose.

I had a case of `assume` pessimizing codegen (will try to create an issue later) on my branch.

**Comment by mikebenfield** [comment]
Interesting. It's certainly the case that LLVM isn't making good use of the `assume`, so maybe I should just get rid of it.

**Comment by nagisa** [comment]
The assumes should definitely be added sparingly as they can pessimize other passes that don't handle it or are forced to preserve it in the patterns which would otherwise end up getting removed outright.

Unless there is a demonstrable improvement in IR quality at various stages of optimization, they should not be output at all.

I’d suggest adding a comment to the assume, documenting those improvements, so that there is a clear pathway to removing them (anybody can check the pattern to see if the assumes are still doing anything).

**Comment by mikebenfield** [comment]
Here's some benchmark data. I benchmarked 3 compilers: `master`, `bgd-assume` (this branch cherry picked onto `master`), and `bgd-noassume` (same but with the `assume` lines commented out).

[Here](https://gist.github.com/mikebenfield/3204ab7ef57f556b4aefb261cdf8f772) is some code for a microbenchmark, which does many matches on a simple enum. 

On my machine (running this like `cargo +master release -- 10000000`), I consistently get around

96ms for master

90ms for bgd-assume

90ms for bgd-noassume

As for benchmarking the compiler itself using the `rustc-perf` tool, here are some instruction counts:
```
('issue-46449', 'Opt-IncrFull')
1,014,347,557 (master)
1,031,204,445 (assume)
988,152,707 (noassume)

('regex', 'Opt-IncrPatched0')
38,185,135,497 (master)
38,336,547,819 (assume)
38,312,812,860 (noassume)

('regex', 'Opt-IncrFull')
88,287,034,172 (master)
88,500,354,715 (assume)
88,457,370,856 (noassume)
```

And here's some CPU time:

```
('issue-46449', 'Opt-IncrFull')
742.517084ms (master)
715.239732ms (assume)
692.373249ms (noassume)

('regex', 'Opt-IncrPatched0')
16.20162392s (master)
15.84068765s (assume)
16.601834155s (noassume)

('regex', 'Opt-IncrFull')
79.921668087s (master)
80.0554681s (assume)
79.500080144s (noassume)
```

I guess based on the instruction counts in particular it does seem best to get rid of the `assume`s. I'll upload a revision with those gone.

**Review by nagisa — COMMENTED** [review]
Inline comments are all largely nits. I haven't been able to find anything partiuclarly wrong with the algorithm, but I would like to see a codegen test (`src/tests/codegen`) added in addition to the UI test that would exercise the three special cases.

I would imagine that the test would basically be a sequence of `CHECK`s in basically the same order as in the implementation. Unfortunately we don’t have the same test case check generator that LLVM does :(

**Comment by nagisa** [comment]
Lets do another perf run now that assumes are gone. I feel like having _some_ regressions with a change as far reaching as this is probably going to be unavoidable. But having some numbers to look at in T-compiler meeting is still necessary, in case there are regressions.

@bors try @rust-timer queue

**Comment by nagisa** [comment]
Looking at the primary benchmarks:

I am failing to find a meaningful difference – the `check` builds regressing might be a second order effect? But actual absolute numbers are so low it is really hard to actually spot any difference.

For `opt` builds the majority of increase is in LTO/optimization passes in LLVM, which I would say is probably expected given that the shape of code has changed slightly and LLVM likely simply has more opportunities to do things instead of giving up. Change in absolute numbers is, again, very small.

The same stands for the big regression in the secondary `issue-46449 | opt | incr-patched: empty 3072` benchmark where the majority of the change seems to be concentrated in LLVM. I believe we can justify this, given this same test in three other scenarios is showing an improvement of a similar degree.

Considering that, and also the improvement shown by microbenchmarks above, I believe that ultimately this is okay to land.

File: compiler/rustc_codegen_ssa/src/mir/place.rs [hunk 6]
```diff
@@ -242,68 +245,156 @@ impl<'a, 'tcx, V: CodegenObject> PlaceRef<'tcx, V> {
...
+                    (tag, t)
                 } else {
-                    bx.sub(tag, bx.cx().const_uint_big(niche_llty, niche_start))
+                    (tag_imm, bx.cx().immediate_backend_type(tag_op.layout))
                 };
+
+                let tag_size = tag_scalar.size(bx.cx());
+                let max_unsigned = tag_size.unsigned_int_max();
+                let max_signed = max_unsigned >> 1;
+                let min_signed = max_signed + 1;
```
[the remark this answers is not in the document]
Comment (mikebenfield) [reply compiler/rustc_codegen_ssa/src/mir/place.rs]: I changed to use `signed_int_max`, but `signed_int_min` does the wrong thing: all the high bits in the `i128` are set.

File: compiler/rustc_codegen_ssa/src/mir/place.rs [hunk 7]
```diff
@@ -242,68 +245,156 @@ impl<'a, 'tcx, V: CodegenObject> PlaceRef<'tcx, V> {
...
-                    bx.icmp(IntPredicate::IntEQ, relative_discr, bx.cx().const_null(niche_llty))
+                let niche_end = niche_start.wrapping_add(relative_max as u128) & max_unsigned;
+                let range = tag_scalar.valid_range(bx.cx());
+
+                let sle = |lhs: u128, rhs: u128| -> bool {
+                    // Toggle the sign bit to do a signed comparison.
+                    (lhs ^ min_signed) <= (rhs ^ min_signed)
+                };
+
+                // We have a subrange `niche_start..=niche_end` inside `range`.
```
[the remark this answers is not in the document]
Comment (mikebenfield) [reply compiler/rustc_codegen_ssa/src/mir/place.rs:277]: If so, I think it should probably happen when a niche is created.

**Comment by mikebenfield** [comment]
What in particular would you like to see in the `codegen` test?

**Comment by nagisa** [comment]
> What in particular would you like to see in the codegen test?

The LLVM-IR for the discriminant reads that exercise the three different cases implemented in this PR, post basic optimization. That way any future changes that would regress (or improve) the codegen here would get properly flagged by the test suite. A regression test of sorts for the referenced issue.

Probably something like this:

```
// compile-flags: -Copt-level=1
// min-llvm-version: 15.0
// ^^^^^^^^^^^^^^^^^^^^^^ if necessary

// CHECK-LABEL: match_case_simple
// CHECK: icmp eq ... or whatever LLVM spits out today
// CHECK-NEXT: ...
// ...
#[no_mangle]
pub fn match_case_simple(enumeration: SimpleEnum) {
    match enumeration { ... } // or std::mem::discriminant()?
}

// ... And similiarly for the other two cases
```

**Comment by mikebenfield** [comment]
Alright, I will add such a test soon.

**Comment by mikebenfield** [comment]
Added a codegen test. I'm just using the names created by LLVM explicitly; let me know if there's a preferabe way.

**Comment by nagisa** [comment]
I think its fine for the time being, even if this isn’t particularly future proof – if it ends up being a pain point, the test can always be adjusted in the future.

@bors r+

**Comment by mikebenfield** [comment]
@nagisa I assume that failure is some network/infrastructure hiccup? Otherwise I'm not sure what's going on.

**Comment by lqd** [comment]
yes

@bors retry network failure in aarch64-gnu builder: "curl: (35) OpenSSL SSL_connect: Connection reset by peer in connection to ci-mirrors.rust-lang.org:443"

**Comment by mikebenfield** [comment]
Alright, that failed because the LLVM IR I test against is not what's generated on 32 bit systems. Duh. I added `// only-x86_64` to the test; someone let me know if that is not an appropriate fix.

**Comment by nagisa** [comment]
It is better to have a `#[no_core]` test that specifies `--target=x86_64-unknown-linux-gnu` or somesuch so that the test can run locally on developers’ non-x86-64 machines, but this is okay too.

@bors r+

**Comment by nnethercote** [comment]
This must have been right on the edge of the regression/no-regression categorization. Not much to worry about here.

@rustbot label: +perf-regression-triaged
